lib/validation.ml

```diff
@@ -339,27 +344,31 @@ let signs pathlen trusted cert =
 let issuer trusted cert =
   List.filter (fun p -> issuer_matches_subject p cert) trusted
 
-let rec validate_anchors revoked pathlen cert = function
+let rec validate_anchors revoked hash pathlen cert = function
   | []    -> Error (`NoTrustAnchor cert)
-  | x::xs -> match signs pathlen x cert with
+  | x::xs -> match signs hash pathlen x cert with
     | Ok _    -> if revoked ~issuer:x ~cert then Error (`Revoked cert) else Ok x
-    | Error _ -> validate_anchors revoked pathlen cert xs
+    | Error _ -> validate_anchors revoked hash pathlen cert xs
 
 let lift_leaf f x =
   match f x with
   | Ok () -> Ok ()
   | Error e -> Error (`Leaf e)
 
-let verify_single_chain ?time ?(revoked = fun ~issuer:_ ~cert:_ -> false) anchors chain =
+let verify_single_chain ?time ?(revoked = fun ~issuer:_ ~cert:_ -> false) hash anchors chain =
   let rec climb pathlen = function
     | cert :: issuer :: certs ->
       is_cert_valid time issuer >>= fun () ->
-      if revoked ~issuer ~cert then Error (`Revoked cert) else Ok () >>= fun () ->
-        signs pathlen issuer cert >>= fun () ->
+      if revoked ~issuer ~cert then
+        Error (`Revoked cert)
+      else begin
+        Ok () >>= fun () ->
+        signs hash pathlen issuer cert >>= fun () ->
```

Comment:
I'm not sure how in what order the checks are done, but it seems that if the hash is not trusted, you don't even need to walk the validation chain (or to do any kind of crypto). Not sure that easier to implement this way, though.

Reply:
that all makes sense, thanks for the writeup!